Cart api with tests

yanazPL opened the pull request:
I've created API for cart.
It includes 3 views for:
1. getting cart content
2. adding items
3. deleting items (all or by quantity)

When deleting:
If no quantity is provided item is deleted.
Otherwise starting quantity of items is decreased by provided quantity.

All views have been tested.

yanazPL committed:
feat: added cart api with tests

yanazPL committed:
fix: minimal conftest changes

jacoor committed:
Merge branch 'dev' into feature/cart-api

jacoor commented on Dshop/apps/products_catalogue/api_urls.py:
```diff
@@ -0,0 +1,10 @@
+
+
+urlpatterns = [
+    path("add/<int:id>/<int:quantity>/", CartAPIView.as_view(), name="api_add_to_cart"),
```
This is not following REST convention. 
REST is using HTTP methods type to do changes, so it should be one item, just different HTTP methods, no?
You already have different, proper handlers in CartAPIView.

yanazPL replied:
done

jacoor commented on Dshop/apps/products_catalogue/tests/test_api_persistent_cart.py:
```diff
@@ -0,0 +1,163 @@
+    assert response.status_code == status.HTTP_200_OK
+    assert "total" in response.data
+    assert "count" in response.data
+    assert "items" in response.data
```
Please add tests for values, not just keys.

yanazPL replied:
done

jacoor commented on Dshop/apps/products_catalogue/tests/test_api_persistent_cart.py:
```diff
@@ -0,0 +1,163 @@
+    client.force_authenticate(user)
+    response = client.post(add_url, {}, content_type="application/json")
+    data = response.data
+    first_item = data.get("items")[0]
```
In tests, IMO, it's better to not use .get. W/o it, it makes test fail cleanly with an exception, here if the key is missing you will not get KeyError but AssertionError. 
Please change. In all places.

yanazPL replied:
done

jacoor commented on Dshop/apps/products_catalogue/tests/test_api_persistent_cart.py:
```diff
@@ -0,0 +1,163 @@
+    response = client.get(reverse("api_cart_detail"), data={}, content_type="application/jspm")
+    data = response.data
+    first_item = data.get("items")[0]
+    assert response.status_code == status.HTTP_200_OK
```
I am thinking, this comparison could be moved to a function since it repeats in at least 2 places. 
ie. def compare_tv_products(data):

yanazPL replied:
done

jacoor reviewed (changes requested):
Nice tests, I like it. Some changes please.

yanazPL committed:
Invalid code posted for correction

yanazPL committed:
Merge branch 'feature/cart-api' of https://github.com/akademiait/Dshop into feature/cart-api

jacoor committed:
Debug and fix serializers

jacoor committed:
ignore

jacoor committed:
small test changes for readability

yanazPL committed:
Fixed conflicts, added CartAPIView with serializers and the first two tests, updated URLs

yanazPL committed:
Merge branch 'dev' of https://github.com/akademiait/Dshop into feature/cart-api

yanazPL committed:
fixed import for pylama

yanazPL committed:
Added error handling, added tests

yanazPL committed:
fix: Corrected repeating test name

jacoor commented on Dshop/apps/products_catalogue/api_views.py:
```diff
@@ -1,12 +1,33 @@
+            read_serializer = CartReadSerializer(cart)
+            return Response(read_serializer.data, status=status.HTTP_201_CREATED)
+        else:
+            print(serializer.errors)
```
Please remove print. It's not a good thing to leave print in code that is production ready.

yanazPL replied:
done

jacoor commented on Dshop/apps/products_catalogue/serializers.py:
```diff
@@ -1,10 +1,67 @@
+        request = self.context['request']
+        cart = get_cart_class().new(request)
+        cart.empty()
+        with transaction.atomic():
```
Actually, can you simply turn on atomic_requests in settings.py? Then this is not necessary.
https://docs.djangoproject.com/en/5.0/topics/db/transactions/#tying-transactions-to-http-requests

yanazPL replied:
imho global only atomic requests would be an overkill. It's probably the only place with `transaction.atomic` in Dshop

jacoor replied:
Your call. But, atomic requests are great thing to tie HTTP request to transaction which makes system much more resilient. Slower, with bigger traffic, but we don't have traffic :P

jacoor commented on Dshop/apps/products_catalogue/tests/test_api_persistent_cart.py:
```diff
@@ -0,0 +1,179 @@
+
+@pytest.mark.django_db
+def test_add(tv_product):
+    client = APIClient()
```
As spoken on discord, please move client, user and force authentication call either to fixture or to function. Fixture preferred, @michalkrzem I think is our expert on fixtures.

yanazPL replied:
done

jacoor reviewed (changes requested):
Small changes please. 
@yanazPL great work, seriously.

jacoor commented on Dshop/apps/products_catalogue/tests/test_api_persistent_cart.py:
```diff
@@ -0,0 +1,179 @@
+    user = User.objects.create_user(username='testuser', password='testpassword')
+    client.force_authenticate(user)
+    product = tv_product
+
```
@yanazPL what does this test do?
I see relogin, but, adding (POST) is before logout so it's not behaving as described. 
It's login, logout, add, check (and not even get, but response check)
Can you please fix this test?

yanazPL replied:
fixed

jacoor commented on Dshop/apps/products_catalogue/tests/test_api_persistent_cart.py:
```diff
@@ -0,0 +1,179 @@
+        ]
+    }
+    response = client.post(reverse("api_cart"), data, format="json")
+    assert response.status_code == status.HTTP_201_CREATED
```
@yanazPL I don't see get request here.  Please add.

yanazPL replied:
done

jacoor commented on Dshop/apps/products_catalogue/tests/test_api_persistent_cart.py:
```diff
@@ -0,0 +1,179 @@
+        ]
+    }
+    response = client.post(reverse("api_cart"), data, format="json")
+    assert response.status_code == status.HTTP_400_BAD_REQUEST
```
Please also add assert for error text. This is good, but it can fail on anything, not only on the error you expect it to.

yanazPL replied:
done

jacoor commented on Dshop/apps/products_catalogue/tests/test_api_persistent_cart.py:
```diff
@@ -0,0 +1,179 @@
+        ]
+    }
+    response = client.post(reverse("api_cart"), data, format="json")
+    assert response.status_code == status.HTTP_400_BAD_REQUEST
```
Missing get call. 
Overall, those tests are only testing POST method, I didn't notice get requests except for test_get_cart_empty and test_get_non_existing_pks.

yanazPL replied:
modified a few post tests to also test get

yanazPL committed:
test: cart api tests refactor

jacoor commented on Dshop/apps/products_catalogue/tests/test_api_persistent_cart.py:
```diff
@@ -0,0 +1,148 @@
+            for product, quantity in zip(ten_tv_products, quantities)
+        ]
+    }
+    response = api_client.post(reverse("api_cart"), data, format="json")
```
One last thing for all tests
please remove format=json.  It's not needed. 
https://github.com/pymasterspl/Dshop/blob/e9bfd0e3087063920873fffc251e0b33304fb04c/Dshop/Dshop/settings.py#L214

yanazPL replied:
I got used to another project without this option set. Will fix it ...

jacoor reviewed (changes requested):
99% there. Great work!

yanazPL committed:
test: removed redundant args in api_client calls

yanazPL committed:
test: added test add_ten_replace_with_one

yanazPL commented:
Removed `format = 'json'` from api_client calls
Removed {} from get api_client calls
Added test apps/products_catalogue/tests/test_api_persistent_cart.py::test_add_ten_replace_with_one

yanazPL committed:
Merge branch 'dev' into feature/cart-api
